File: apps/web/src/lib/evaluation.ts
import { apiGet } from "./api";

/**
 * Data evaluasi Prediction vs Actual (TASK 150–151).
 *
 * Menopang success criteria #06 Taskap. Setiap bentuk data di sini membawa `status`
 * dan `basis` dari API, dan keduanya **wajib ikut tampil**: aturan pencocokan antara
 * prediksi dan kejadian nyata belum ditetapkan (U-03), sehingga angka precision/recall
 * belum boleh dibaca sebagai hasil final (CLAUDE.md §26).
 */

export type EvaluationMetrics = {
  hits: number;
  false_positives: number;
  false_negatives: number;
  /** `null` berarti tidak dapat dihitung — berbeda maknanya dari nol. */
  precision: number | null;
  recall: number | null;
  evaluated_rows: number;
  status: string;
  basis: string;
};

/** Hasil pencocokan satu baris evaluasi. */
export type MatchType = "HIT" | "FALSE_POSITIVE" | "FALSE_NEGATIVE";

export type EvaluationSummary = {
  /** Jenis ancaman → jumlah baris per hasil pencocokan. */
  per_threat: Record<string, Partial<Record<MatchType, number>>>;
  model_versions: string[];
  status: string;
  basis: string;
};

export const MATCH_TYPES: readonly MatchType[] = [
  "HIT",
  "FALSE_POSITIVE",
  "FALSE_NEGATIVE",
] as const;

export const MATCH_LABELS: Record<MatchType, string> = {
  HIT: "Terbukti",
  FALSE_POSITIVE: "Positif Palsu",
  FALSE_NEGATIVE: "Negatif Palsu",
};

export const MATCH_HINTS: Record<MatchType, string> = {
  HIT: "Diprediksi dan benar-benar terjadi",
  FALSE_POSITIVE: "Diprediksi, tetapi tidak terjadi",
  FALSE_NEGATIVE: "Terjadi, tetapi tidak diprediksi",
};

/** Angka rasio dalam gaya Indonesia (koma desimal). Nilai kosong dinyatakan, bukan diisi nol. */
export function formatRatio(value: number | null): string {
  if (value === null) return "tidak dapat dihitung";
  return value.toLocaleString("id-ID", { minimumFractionDigits: 3, maximumFractionDigits: 3 });
}

/** Bentuk persen untuk pembaca non-teknis. */
export function formatPercent(value: number | null): string | null {
  if (value === null) return null;
  return `${(value * 100).toLocaleString("id-ID", { maximumFractionDigits: 1 })}%`;
}

export const getEvaluationMetrics = () => apiGet<EvaluationMetrics>("/evaluation/metrics");

export const getEvaluationSummary = () => apiGet<EvaluationSummary>("/evaluation/summary");

File: apps/api/src/prediksi_presisi_api/api/routers/evaluation.py
"""API evaluasi model (TASK 040).

Menopang success criteria #06 Taskap: **Prediction vs Actual**.

Angka precision/recall di sini **selalu** disertai penanda `PROPOSED` dan penjelasan
cakupannya. Aturan pencocokan spasial-temporal antara prediksi dan kejadian nyata belum
ditetapkan (U-03), sehingga menyajikannya sebagai angka final akan melampaui yang benar-benar
dapat dipertanggungjawabkan (docs/05 §2.10, CLAUDE.md §26).
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from ...models import Prediction, PredictionActual
from ...seeding.regenerate import EVALUATED_THREATS
from ..deps import CurrentUser, get_db, require_permission

router = APIRouter(prefix="/evaluation", tags=["evaluasi"])

BASIS = (
    "Cakupan: kejadian nyata pada periode prediksi untuk jenis ancaman yang diprediksi "
    f"({', '.join(EVALUATED_THREATS)}). False negative dihitung dari kejadian pada sel "
    "tanpa prediksi terbit. Aturan pencocokan final belum ditetapkan (U-03)."
)


def _counts(session: Session) -> dict[str, int]:
    rows = session.execute(
        select(PredictionActual.match_type, func.count()).group_by(PredictionActual.match_type)
    ).all()
    return {match_type: total for match_type, total in rows}


@router.get("/metrics", summary="Precision, recall, FP, FN")
def metrics(
    session: Session = Depends(get_db),
    current: CurrentUser = require_permission("evaluation:read"),
) -> dict[str, Any]:
    counts = _counts(session)
    hits = counts.get("HIT", 0)
    false_positives = counts.get("FALSE_POSITIVE", 0)
    false_negatives = counts.get("FALSE_NEGATIVE", 0)

    predicted = hits + false_positives
    actual = hits + false_negatives

    return {
        "hits": hits,
        "false_positives": false_positives,
        "false_negatives": false_negatives,
        # Pembagian nol dijawab None, bukan 0: "tidak dapat dihitung" berbeda maknanya
        # dari "nilainya nol".
        "precision": round(hits / predicted, 3) if predicted else None,
        "recall": round(hits / actual, 3) if actual else None,
        "evaluated_rows": sum(counts.values()),
        "status": "PROPOSED",
        "basis": BASIS,
    }


@router.get("/summary", summary="Ringkasan evaluasi per jenis ancaman")
def summary(
    session: Session = Depends(get_db),
    current: CurrentUser = require_permission("evaluation:read"),
) -> dict[str, Any]:
    rows = session.execute(
        select(
            PredictionActual.actual_threat_type,
            PredictionActual.match_type,
            func.count(),
        ).group_by(PredictionActual.actual_threat_type, PredictionActual.match_type)
    ).all()

    per_threat: dict[str, dict[str, int]] = {}
    for threat, match_type, total in rows:
        bucket = per_threat.setdefault(threat or "TIDAK DIKETAHUI", {})
        bucket[match_type] = total

    evaluated_models = session.scalars(
        select(Prediction.model_version).distinct().order_by(Prediction.model_version)
    ).all()

    return {
        "per_threat": per_threat,
        "model_versions": list(evaluated_models),
        "status": "PROPOSED",
        "basis": BASIS,
    }
